File: src/main/java/org/cubexmc/model/PowerCondition.java
package org.cubexmc.model;

import org.bukkit.World;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;

/**
 * 权力条件配置
 * 定义权力结构生效的条件（时间、世界等）
 * 可用于宝石和任命的条件判断
 */
public class PowerCondition {

    // 时间条件
    private boolean timeEnabled = false;
    private TimeType timeType = TimeType.ALWAYS;
    private long timeFrom = 0;
    private long timeTo = 24000;

    // 世界条件
    private boolean worldEnabled = false;
    private WorldMode worldMode = WorldMode.WHITELIST;
    private List<String> worldList = new ArrayList<>();

    /**
     * 时间类型枚举
     */
    public enum TimeType {
        ALWAYS,   // 始终生效
        DAY,      // 白天 (0-12000)
        NIGHT,    // 夜晚 (12000-24000)
        CUSTOM    // 自定义时间范围
    }

    /**
     * 世界模式枚举
     */
    public enum WorldMode {
        WHITELIST,  // 白名单模式：只在列表中的世界生效
        BLACKLIST   // 黑名单模式：在列表外的世界生效
    }

    /**
     * 检查玩家是否满足所有条件
     * @param player 玩家
     * @return 是否满足条件
     */
    public boolean checkConditions(Player player) {
        if (player == null || !player.isOnline()) return false;
        
        // 检查时间条件
        if (!checkTimeCondition(player)) return false;
        
        // 检查世界条件
        if (!checkWorldCondition(player)) return false;
        
        return true;
    }

    /**
     * 检查时间条件
     */
    public boolean checkTimeCondition(Player player) {
        if (!timeEnabled) return true;
        
        World world = player.getWorld();
        long time = world.getTime(); // 0-24000
        
        switch (timeType) {
            case ALWAYS:
                return true;
            case DAY:
                // 白天: 0-12000 (日出到日落)
                return time >= 0 && time < 12000;
            case NIGHT:
                // 夜晚: 12000-24000 (日落到日出)
                return time >= 12000 && time < 24000;
            case CUSTOM:
                // 自定义时间范围
                if (timeFrom <= timeTo) {
                    // 正常范围 (如 6000-18000)
                    return time >= timeFrom && time < timeTo;
                } else {
                    // 跨越午夜的范围 (如 18000-6000)
                    return time >= timeFrom || time < timeTo;
                }
            default:
                return true;
        }
    }

    /**
     * 检查世界条件
     */
    public boolean checkWorldCondition(Player player) {
        if (!worldEnabled || worldList.isEmpty()) return true;
        
        String worldName = player.getWorld().getName();
        boolean inList = worldList.contains(worldName);
        
        switch (worldMode) {
            case WHITELIST:
                return inList;
            case BLACKLIST:
                return !inList;
            default:
                return true;
        }
    }

    /**
     * 获取条件描述（用于显示）
     */
    public String getConditionDescription() {
        StringBuilder sb = new StringBuilder();
        
        if (timeEnabled && timeType != TimeType.ALWAYS) {
            switch (timeType) {
                case DAY:
                    sb.append("白天生效");
                    break;
                case NIGHT:
                    sb.append("夜晚生效");
                    break;
                case CUSTOM:
                    sb.append("时间 ").append(timeFrom).append("-").append(timeTo);
                    break;
                case ALWAYS:
                default:
                    // 无时间限制
                    break;
            }
        }
        
        if (worldEnabled && !worldList.isEmpty()) {
            if (sb.length() > 0) sb.append(", ");
            sb.append(worldMode == WorldMode.WHITELIST ? "仅在: " : "排除: ");
            sb.append(String.join(", ", worldList));
        }
        
        return sb.length() > 0 ? sb.toString() : "无限制";
    }

    /**
     * 是否有任何条件启用
     */
    public boolean hasAnyCondition() {
        return (timeEnabled && timeType != TimeType.ALWAYS) || 
               (worldEnabled && !worldList.isEmpty());
    }

    /**
     * 创建副本
     */
    public PowerCondition copy() {
        PowerCondition copy = new PowerCondition();
        copy.timeEnabled = this.timeEnabled;
        copy.timeType = this.timeType;
        copy.timeFrom = this.timeFrom;
        copy.timeTo = this.timeTo;
        copy.worldEnabled = this.worldEnabled;
        copy.worldMode = this.worldMode;
        copy.worldList = new ArrayList<>(this.worldList);
        return copy;
    }

    // ==================== Getters and Setters ====================

    public boolean isTimeEnabled() {
        return timeEnabled;
    }

    public void setTimeEnabled(boolean timeEnabled) {
        this.timeEnabled = timeEnabled;
    }

    public TimeType getTimeType() {
        return timeType;
    }

    public void setTimeType(TimeType timeType) {
        this.timeType = timeType;
    }

    public long getTimeFrom() {
        return timeFrom;
    }

    public void setTimeFrom(long timeFrom) {
        this.timeFrom = timeFrom;
    }

    public long getTimeTo() {
        return timeTo;
    }

    public void setTimeTo(long timeTo) {
        this.timeTo = timeTo;
    }

    public boolean isWorldEnabled() {
        return worldEnabled;
    }

    public void setWorldEnabled(boolean worldEnabled) {
        this.worldEnabled = worldEnabled;
    }

    public WorldMode getWorldMode() {
        return worldMode;
    }

    public void setWorldMode(WorldMode worldMode) {
        this.worldMode = worldMode;
    }

    public List<String> getWorldList() {
        return worldList;
    }

    public void setWorldList(List<String> worldList) {
        this.worldList = worldList != null ? worldList : new ArrayList<>();
    }
}

File: src/main/java/org/cubexmc/model/PowerCondition.kt
package org.cubexmc.model

import org.bukkit.entity.Player
import org.cubexmc.manager.LanguageManager

/**
 * 权力条件配置。
 */
class PowerCondition {
    var isTimeEnabled: Boolean = false
    var timeType: TimeType = TimeType.ALWAYS
        set(value) {
            field = value
        }
    var timeFrom: Long = 0
    var timeTo: Long = 24000
    var isWorldEnabled: Boolean = false
    var worldMode: WorldMode = WorldMode.WHITELIST
        set(value) {
            field = value
        }
    var worldList: List<String> = ArrayList()
        set(value) {
            field = value ?: ArrayList()
        }

    enum class TimeType {
        ALWAYS,
        DAY,
        NIGHT,
        CUSTOM,
    }

    enum class WorldMode {
        WHITELIST,
        BLACKLIST,
    }

    fun checkConditions(player: Player?): Boolean {
        if (player == null || !player.isOnline) return false
        if (!checkTimeCondition(player)) return false
        if (!checkWorldCondition(player)) return false
        return true
    }

    fun checkTimeCondition(player: Player): Boolean {
        if (!isTimeEnabled) return true
        val time = player.world.time
        return when (timeType) {
            TimeType.ALWAYS -> true
            TimeType.DAY -> time in 0 until 12000
            TimeType.NIGHT -> time in 12000 until 24000
            TimeType.CUSTOM -> if (timeFrom <= timeTo) {
                time >= timeFrom && time < timeTo
            } else {
                time >= timeFrom || time < timeTo
            }
        }
    }

    fun checkWorldCondition(player: Player): Boolean {
        if (!isWorldEnabled || worldList.isEmpty()) return true
        val worldName = player.world.name
        val inList = worldList.contains(worldName)
        return when (worldMode) {
            WorldMode.WHITELIST -> inList
            WorldMode.BLACKLIST -> !inList
        }
    }

    @JvmOverloads
    fun getConditionDescription(lang: LanguageManager? = null): String {
        val builder = StringBuilder()
        if (isTimeEnabled && timeType != TimeType.ALWAYS) {
            when (timeType) {
                TimeType.DAY -> builder.append(msg(lang, "condition.time_day", "Day only"))
                TimeType.NIGHT -> builder.append(msg(lang, "condition.time_night", "Night only"))
                TimeType.CUSTOM -> builder.append(msg(lang, "condition.time_custom", "Time")).append(" ").append(timeFrom).append("-").append(timeTo)
                TimeType.ALWAYS -> {}
            }
        }

        if (isWorldEnabled && worldList.isNotEmpty()) {
            if (builder.isNotEmpty()) builder.append(", ")
            builder.append(
                if (worldMode == WorldMode.WHITELIST) {
                    msg(lang, "condition.world_whitelist", "Only in: ")
                } else {
                    msg(lang, "condition.world_blacklist", "Except: ")
                },
            )
            builder.append(worldList.joinToString(", "))
        }

        return if (builder.isNotEmpty()) builder.toString() else msg(lang, "condition.no_limit", "No restrictions")
    }

    fun hasAnyCondition(): Boolean = (isTimeEnabled && timeType != TimeType.ALWAYS) || (isWorldEnabled && worldList.isNotEmpty())

    fun copy(): PowerCondition {
        val copy = PowerCondition()
        copy.isTimeEnabled = isTimeEnabled
        copy.timeType = timeType
        copy.timeFrom = timeFrom
        copy.timeTo = timeTo
        copy.isWorldEnabled = isWorldEnabled
        copy.worldMode = worldMode
        copy.worldList = ArrayList(worldList)
        return copy
    }

    companion object {
        private fun msg(lang: LanguageManager?, key: String, fallback: String): String {
            if (lang == null) return fallback
            val value = lang.getMessage(key)
            return if (value != null && !value.startsWith("Missing message")) value else fallback
        }
    }
}
